Active Record Query Interface > 18. Find or Build a New Object (#363)
campuscode/rails-guides-pt-BR: pt-BR/active_record_querying.md

COMMIT
```

`find_or_create_by` retorna o registro que já existe ou o novo registro. Em nosso caso, ainda não tínhamos um cliente chamado Andy, então o registro é criado e retornado.

O novo registro pode não ser salvo no banco de dados; isso depende se as validações foram aprovadas ou não (assim como `create`).

> Suponha que queremos definir o atributo 'bloqueado' para `false` se estamos

Author: :top:

criando um novo registro, mas não queremos incluí-lo na consulta. Então
queremos encontrar o cliente chamado "Andy", ou se esse cliente não

> existir, crie um cliente chamado "Andy" que não esteja bloqueado.

Reviewer: ```suggestion
existir, crie um cliente chamado "Andy" que não esteja bloqueado.
```

Podemos conseguir isso de duas maneiras. A primeira é usar `create_with`:

```ruby
Client.create_with(locked: false).find_or_create_by(first_name: 'Andy')
```

> A segunda maneira é usar um _lock_:

Author: Eu fiquei muito em dúvida, porque ele faz a busca por um bloco de código, mas ao mesmo tempo usa um lock...
Parece que ambos fazem sentido rs

> A segunda maneira é usar um _lock_:

Reviewer 2: > Eu fiquei muito em dúvida, porque ele faz a busca por um bloco de código, mas ao mesmo tempo usa um lock...
> Parece que ambos fazem sentido rs

Pela continuação do parágrafo logo abaixo, trata-se de um "bloco" mesmo.

### `find_or_initialize_by`

> O `find_or_initialize_by` método funcionará como o

Author: ```suggestion
O método `find_or_initialize_by` funcionará como o
```

> O `find_or_initialize_by` método funcionará como o

Reviewer: Essa aqui você mudou enquanto eu fazia a revisão então já fechei 😅

> `find_or_create_by` mas irá chamar `new` ao invés de `create`. Isso significa que uma nova instância de modelo será criada na memória, mas não será salva no banco de dados. Continuando com o exemplo `find_or_create_by`, agora queremos o cliente chamado 'Nick':

Reviewer 2: ```suggestion
`find_or_create_by` mas chamará `new` ao invés de `create`. Isso significa que uma nova instância do *model* será criada na memória, mas não será salva no banco de dados. Continuando com o exemplo `find_or_create_by`, agora queremos o cliente chamado 'Nick':
```

Author: Perfeito! Falhei na convenção nesse parágrafo :facepalm:
